core/src/Cache/Keys/MsalIdTokenCacheKey.cs

```diff
@@ -0,0 +1,140 @@
+
+using System;
+using System.Text;
+
+namespace Microsoft.Identity.Core.Cache
+{
+    /// <summary>
+    /// An object representing the key of the token cache Id Token dictionary. The 
+    /// format of the key is not important for this library, as long as it is unique.
+    /// </summary>
+    internal class MsalIdTokenCacheKey
+    {
+        private readonly string _environment;
+        private readonly string _homeAccountId;
+        private readonly string _clientId;
+        private readonly string _tenantId;
+
+        public MsalIdTokenCacheKey(
+            string environment,
+            string tenantId,
+            string userIdentifier,
+            string clientId)
+
+        {
+            if (string.IsNullOrEmpty(environment))
+            {
+                throw new ArgumentNullException(nameof(environment));
+            }
+
+            if (string.IsNullOrEmpty(clientId))
+            {
+                throw new ArgumentNullException(nameof(clientId));
+            }
+
+            _environment = environment;
+            _homeAccountId = userIdentifier;
+            _clientId = clientId;
+            _tenantId = tenantId;
+        }
+
+
+        public override string ToString()
+        {
+            var stringBuilder = new StringBuilder();
+
+            stringBuilder.Append(_homeAccountId ?? "");
+            stringBuilder.Append(MsalCacheConstants.CacheKeyDelimiter);
+
+            stringBuilder.Append(_environment);
+            stringBuilder.Append(MsalCacheConstants.CacheKeyDelimiter);
+
+            stringBuilder.Append(MsalCacheConstants.IdToken);
+            stringBuilder.Append(MsalCacheConstants.CacheKeyDelimiter);
+
+            stringBuilder.Append(_clientId);
+            stringBuilder.Append(MsalCacheConstants.CacheKeyDelimiter);
+
+            stringBuilder.Append(_tenantId ?? "");
+            stringBuilder.Append(MsalCacheConstants.CacheKeyDelimiter);
+
+            return stringBuilder.ToString();
+        }
+
+        #region iOS
+
+        public string GetiOSAccountKey()
+        {
+            var stringBuilder = new StringBuilder();
+
+            stringBuilder.Append(_homeAccountId ?? "");
+            stringBuilder.Append(MsalCacheConstants.CacheKeyDelimiter);
+
+            stringBuilder.Append(_environment);
+
+            return stringBuilder.ToString();
+        }
+
+        public string GetiOSServiceKey()
+        {
+            var stringBuilder = new StringBuilder();
+
+            stringBuilder.Append(MsalCacheConstants.IdToken);
+            stringBuilder.Append(MsalCacheConstants.CacheKeyDelimiter);
+
+            stringBuilder.Append(_clientId);
+            stringBuilder.Append(MsalCacheConstants.CacheKeyDelimiter);
+
+            stringBuilder.Append(_tenantId ?? "");
+            stringBuilder.Append(MsalCacheConstants.CacheKeyDelimiter);
+
+            return stringBuilder.ToString();
+        }
+
+        public string GetiOSGenericKey()
```

Comment:
same for credentials - consider moving iOS type into here too for consistency

Reply:
That type seems to be used in many places and the current implementation (an enum) is pretty readable as is. 

---
In reply to: [221722252](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1287#discussion_r221722252) [](ancestors = 221722252)